# Import Libraries and Dataframes

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create path
path = "/Users/charlottelin/Documents/06-2025 Instacart Basket Analysis"

In [3]:
# Import ords_prods_merge with extra columns
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_extra_columns.pkl'))

In [4]:
# Check import
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

   Unnamed: 0_y                             product_name  aisle_id  \
0           195                                     Soda        77   
1         14084  Organic Unsweetened Vanilla Almond Milk        91   
2         12427                      Original Beef Jerky        23   
3         26089               Aged White Cheddar Popcorn        23   
4         26406         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match    price_range_loc     busiest_day  \
0              7     9.0  both  Mid-range product  Regularly busy   
1             16    12.5  both  Mid-range product  Regularly busy   
2             19     4.4  both  Low-range product  Regularly busy   
3             19     4.7  both  Low-range product  Regularly busy   
4             17     1.0  both  Low-range product  Regularly busy   

   busiest_days busiest_period_of_day  
0  Regular days        Average Orders  
1  Regular days        Average Orders  
2  Regular days        Average Orders  
3  Regular days        Average Orders  
4  Regular days        Average Orders

In [5]:
# Drop extra unnamed column
ords_prods_merge = ords_prods_merge.drop(columns=['Unnamed: 0_y'])

In [6]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  match    price_range_loc     busiest_day  busiest_days busiest_period_of_day  
0  both  Mid-range product  Regularly busy  Regular days        Average Orders  
1  both  Mid-range product  Regularly busy  Regular days        Average Orders  
2  both  Low-range product  Regularly busy  Regular days        Average Orders  
3  both  Low-range product  Regularly busy  Regular days        Average Orders  
4  both  Low-range product  Regularly busy  Regular days        Average Orders

# Create a subset to manage memory

In [7]:
df = ords_prods_merge[:100000]

In [8]:
df.shape

(100000, 19)

# Grouping & Aggregating Data

In [9]:
df.groupby('product_name')

### Aggregate mean of "order_number" by department_id to the subset

In [10]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.256635
2                14.508929
3                14.803771
4                17.720406
5                21.806678
6                14.230856
7                17.532278
8                18.044481
9                16.422806
10               12.072581
11               14.774628
12               14.127996
13               16.187101
14               17.527778
15               15.913782
16               17.462443
17               15.905701
18               25.801959
19               17.310490
20               16.232737
21               21.155738

In [11]:
df.groupby('department_id')['order_number'].mean()

department_id
1     14.256635
2     14.508929
3     14.803771
4     17.720406
5     21.806678
6     14.230856
7     17.532278
8     18.044481
9     16.422806
10    12.072581
11    14.774628
12    14.127996
13    16.187101
14    17.527778
15    15.913782
16    17.462443
17    15.905701
18    25.801959
19    17.310490
20    16.232737
21    21.155738
Name: order_number, dtype: float64

In [12]:
# Adding multiple stats to order_number
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.256635   1  98
2                14.508929   1  98
3                14.803771   1  99
4                17.720406   1  99
5                21.806678   1  99
6                14.230856   1  97
7                17.532278   1  99
8                18.044481   1  51
9                16.422806   1  99
10               12.072581   1  55
11               14.774628   1  94
12               14.127996   1  98
13               16.187101   1  99
14               17.527778   1  99
15               15.913782   1  99
16               17.462443   1  99
17               15.905701   1  95
18               25.801959   1  95
19               17.310490   1  99
20               16.232737   1  98
21               21.155738   1  93

### Step 2. Aggregate mean of "order_number" column grouped by "department_id" for the entire dataframe

In [13]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

### Step 3. Compare results of the entire dataframe vs the subset

#### Comparison Insights
- The average `order_number` by `department_id` in the subset and full dataset are closely aligned, showing some, but minimal variation.
- Over half of the departments have identical or near-identical mean values between the two datasets.
- A few departments show modest differences, likely due to sample size variability or less frequent ordering patterns.
- Overall, the subset of 1 million rows appears to be a statistically representative sample of the full dataset.

### Step 4. Create a loyalty flag for existing customers

In [14]:
# Create a new 'max_order' column, split the data into 'user_id' column, and apply transform() function to generate the maximum orders for each user.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_52382/3180993555.py:2: FutureWarning: The provided callable <function max at 0x104246520> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [18]:
# Check results
ords_prods_merge[['user_id', 'order_number', 'max_order']].head(25)

user_id  order_number  max_order
0         1             1         10
1         1             1         10
2         1             1         10
3         1             1         10
4         1             1         10
5         1             2         10
6         1             2         10
7         1             2         10
8         1             2         10
9         1             2         10
10        1             2         10
11        1             3         10
12        1             3         10
13        1             3         10
14        1             3         10
15        1             3         10
16        1             4         10
17        1             4         10
18        1             4         10
19        1             4         10
20        1             4         10
21        1             5         10
22        1             5         10
23        1             5         10
24        1             5         10

#### The above list shows all User IDs of 1, which may be due to the limited sample size. Therefore, I will run another code to look at a more random sample from the entire dataframe.

In [19]:
ords_prods_merge[['user_id', 'order_number', 'max_order']].sample(25)

user_id  order_number  max_order
7910058     50187             4         12
9168171     58232             3          3
14029994    88935             3          9
27715205   176289             4          6
16489535   104412             7         14
13337448    84467             9         12
26782280   170453             8         19
28451485   180986            13         30
3824220     24304             1          8
20756861   132006            44         81
22051280   140192            37         68
2849896     18224             4         68
11369647    72110            53         73
21317148   135516             2          3
18422860   116698            78         89
18189016   115231             1          5
1131983      7326            26         44
4325142     27514            11         47
3708502     23577            25         30
17478799   110665             4          7
1439060      9291            37         50
20293970   128933             1          6
13593173    86120            25         31
32168945   204652             5         12
18524978   117306            18         18

#### Apply filters to flag

In [21]:
# 'Loyal customer' > 40 orders; 'Regular customer' = 11-40 orders; 'New customer' < 11 orders.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

#### Check frequency

In [22]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

#### Check output
- I'm using sample() again to broaden the list

In [24]:
ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].sample(25)

user_id  max_order      loyalty_flag
24788402   157517          6      New customer
3938004     25065          5      New customer
30802939   195855          5      New customer
27295643   173643         26  Regular customer
7414540     47074          4      New customer
14241492    90251         23  Regular customer
26784208   170467         41    Loyal customer
25110273   159567         67    Loyal customer
31439428   199960         84    Loyal customer
18647737   118134         49    Loyal customer
7702634     48935         20  Regular customer
4924563     31391         36  Regular customer
3220824     20719          8      New customer
8658082     54890         16  Regular customer
14509032    91925         43    Loyal customer
14444855    91534          6      New customer
18258617   115681         41    Loyal customer
11568361    73338         39  Regular customer
5832594     37169          8      New customer
18937687   120010         16  Regular customer
8582475     54408         58    Loyal customer
30280542   192573          9      New customer
15694111    99413         18  Regular customer
10959135    69505         31  Regular customer
1548780     10033         10      New customer

## Step 5. Insights for Marketing: customers spending habits

In [25]:
# Check the basic stats per loyalty category
ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count      mean       std  min  25%  50%   75%   max
loyalty_flag                                                               
Loyal customer    10284093.0  7.773487  4.205886  1.0  4.2  7.4  11.2  25.0
New customer       6243990.0  7.801077  4.266452  1.0  4.2  7.4  11.3  25.0
Regular customer  15876776.0  7.798145  4.254540  1.0  4.2  7.4  11.3  25.0

In [26]:
# Check the median too to make sure the mean is not skewed
ords_prods_merge.groupby('loyalty_flag')['prices'].median()

loyalty_flag
Loyal customer      7.4
New customer        7.4
Regular customer    7.4
Name: prices, dtype: float64

#### <b>Conclusion:</b> 
There is overlal very little variation in average prices between loyalty categories. The mean price for all segments is around $7.77–$7.80, and the median is consistently $7.4. All customer types share similar price distributions, with identical min, quartile, and max values. <b>This suggests that loyalty status does not significantly affect spending per product.</b>


## Step 6. Target diffrent types of spenders

In [28]:
# Create a spending flag with a new 'avg_spend_user' column
ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_52382/324571645.py:2: FutureWarning: The provided callable <function mean at 0x104246f20> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [30]:
# Check results
ords_prods_merge[['user_id', 'prices', 'avg_spend_user']].sample(25)

user_id  prices  avg_spend_user
12859096    81430    11.9        8.450000
28084553   178660     4.5        7.055797
14206910    90024     6.3        6.790244
16797544   106421     6.6        8.091186
13882841    87984    15.0        7.702151
25697100   163322     6.8        7.753533
31729058   201827    11.7        8.607308
1346835      8697    14.6        8.344048
2630153     16843     8.2        7.744325
3490201     22346     9.2        6.872462
9558878     60716    13.4        8.319915
25226972   160292    12.1        7.836311
20439290   129846    13.7        8.233000
5212076     33154     6.5        6.607463
17585859   111302     1.3        7.766160
24680315   156808    13.9        7.157692
20421944   129747    13.6        7.348092
10758940    68282     4.5        7.121749
12031864    76238    10.2        8.061864
10168004    64752     2.0        7.259369
2075339     13335    11.4        6.853654
21728485   138214    13.2        8.866812
2062916     13253     4.3        8.446259
4982400     31760     1.2        7.528290
20495932   130204    12.2        7.117371

In [31]:
# Apply filter to the new 'avg_spend_user' column: 'Low spender' < 10; 'High spender' >= 10.
ords_prods_merge.loc[ords_prods_merge['avg_spend_user'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_spend_user'] >= 10, 'spending_flag'] = 'High spender'

In [32]:
# Check frequency
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32285150
High spender      119709
Name: count, dtype: int64

In [33]:
# Check output
ords_prods_merge[['user_id', 'avg_spend_user', 'spending_flag']].sample(25)

user_id  avg_spend_user spending_flag
18528399   117337        9.417857   Low spender
26581230   169165        7.823980   Low spender
22460499   142763        9.277083   Low spender
7081276     44985        8.404651   Low spender
30092325   191383        7.911921   Low spender
1011969      6551        7.240379   Low spender
4032060     25610        8.006297   Low spender
17589045   111330        8.193355   Low spender
6683891     42550        7.460908   Low spender
6524020     41553        7.559524   Low spender
15527804    98315        7.096535   Low spender
26078943   165853        6.235897   Low spender
23262661   147845        9.052727   Low spender
21224068   134964       10.587719  High spender
19048459   120746        7.300455   Low spender
9453205     60078        8.859416   Low spender
31379974   199599        7.547920   Low spender
29352440   186587        7.568750   Low spender
11914826    75498        8.043820   Low spender
11313461    71758        8.263953   Low spender
10278665    65441        7.076404   Low spender
11864226    75168        8.177728   Low spender
22324570   141916        8.974519   Low spender
1337781      8651        8.946468   Low spender
22471536   142853        7.465077   Low spender

## Step 7. Send relevant notifications to users by order frequency

In [34]:
# Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.
ords_prods_merge['ord_regularity_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_52382/679455541.py:2: FutureWarning: The provided callable <function median at 0x104381bc0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['ord_regularity_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [35]:
# Check results 
ords_prods_merge[['user_id', 'days_since_prior_order', 'ord_regularity_median']].sample(25)

user_id  days_since_prior_order  ord_regularity_median
8052049     51118                     7.0                    8.0
7326145     46455                    17.0                    9.0
6818359     43352                     1.0                    2.0
15035179    95238                    16.0                    6.0
21104239   134266                     8.0                   12.0
9018747     57185                     9.0                    9.0
27071530   172235                    17.0                   18.0
13038528    82566                     8.0                    7.0
2555344     16384                    11.0                   10.0
369726       2456                     6.0                    5.0
17257747   109338                     NaN                   30.0
29564258   187973                    22.0                   30.0
17175687   108790                     2.0                    6.0
20239282   128548                    20.0                   15.0
6570616     41862                     4.0                    5.0
11273271    71489                     9.0                    6.0
15807713   100108                     6.0                    4.0
9189555     58406                     2.0                    3.0
4845408     30916                    30.0                   14.0
27390882   174268                     7.0                    6.0
29271118   186076                    28.0                   28.0
15634097    99019                     5.0                    6.0
18867915   119516                     NaN                   12.0
6558121     41788                    12.0                    8.0
13025116    82505                    16.0                   17.0

In [36]:
# Apply filters to 'ord_regularity_median" column: 'Non-frequent customer' > 20; 'Regular customer' = 10-20; 'Frequent customer' <= 10.
ords_prods_merge.loc[ords_prods_merge['ord_regularity_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['ord_regularity_median'] > 10) & (ords_prods_merge['ord_regularity_median'] <= 20), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['ord_regularity_median'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [37]:
# Check results
ords_prods_merge[['user_id', 'days_since_prior_order', 'ord_regularity_median', 'frequency_flag']].sample(25)

user_id  days_since_prior_order  ord_regularity_median  \
4400405     28002                    12.0                   13.0   
5433569     34565                     2.0                    5.0   
24509063   155775                     NaN                    5.0   
14477321    91741                     8.0                   10.0   
24525308   155863                     7.0                    5.0   
25637004   162950                    15.0                    5.0   
8902049     56389                     6.0                    6.0   
6715163     42738                     6.0                    7.0   
17321590   109747                    11.0                   11.0   
1364219      8805                     4.0                    6.0   
12531968    79345                     7.0                    7.0   
10468257    66535                    23.0                   23.0   
5691711     36272                     2.0                    4.0   
21716256   138149                     0.0                    7.0   
2242521     14435                     6.0                    8.0   
25436496   161687                     9.0                    5.0   
23837486   151494                    13.0                   13.0   
17574732   111225                    30.0                   14.0   
22968312   145990                    30.0                   23.0   
4831865     30846                     NaN                   13.0   
26186947   166573                     7.0                    7.0   
5931706     37809                     5.0                    6.0   
10251969    65293                     7.0                   10.5   
5528543     35180                    14.0                   12.0   
10381452    66090                     3.0                    4.0   

                 frequency_flag  
4400405        Regular customer  
5433569       Frequent customer  
24509063      Frequent customer  
14477321      Frequent customer  
24525308      Frequent customer  
25637004      Frequent customer  
8902049       Frequent customer  
6715163       Frequent customer  
17321590       Regular customer  
1364219       Frequent customer  
12531968      Frequent customer  
10468257  Non-frequent customer  
5691711       Frequent customer  
21716256      Frequent customer  
2242521       Frequent customer  
25436496      Frequent customer  
23837486       Regular customer  
17574732       Regular customer  
22968312  Non-frequent customer  
4831865        Regular customer  
26186947      Frequent customer  
5931706       Frequent customer  
10251969       Regular customer  
5528543        Regular customer  
10381452      Frequent customer

In [38]:
# Check frequency
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [40]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False).sum()

32404859

## Export final pickle

In [41]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4_8.pkl'))